### ISSUE TITLE
Training a multi-class model for WM, GM and lesion

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star#2]
WM and GM are mutually exclusive, whereas lesions could be part of the WM and/or the GM class.

In this case, what would be the best approach for training a model?

- Option 1: Train two separate models: WM/GM and lesion
- Option 2: Train a single model, with multiple classes, some that could overlap (lesion) and some that could not (WM, GM).

**Comment by jcohenadad** [comment]
Thank you @plbenveniste! And how about training a SC-GM-lesion model instead of WM-GM-lesion model? In the first one, the SC and GM are no more mutually exclusive.

**Comment by plbenveniste** [comment]
Yes, it would make sense defining the following regions: 
```json
"labels": {
    "background": 0,
    "spinal_cord": [1, 2, 3],
    "grey_matter": [2],
    "lesions": [3]
},
"regions_class_order": [1, 2, 3],
```

Also, I agree with you: I think using SC-GM-Lesion would make more sense than SC-WM-Lesion, as WM would be considered SC minus GM. This choice would avoid having a contour of GM in areas where the model would not be 100% confident.

**Comment by naga-karthik** [comment]
Just adding my 2c here: I agree with @plbenveniste about training the SC-GM-Lesion model instead of the SC-WM-lesion. Region-based approach is ideal precisely for the SC-GM-lesion scenario where the smaller objects are the subset of the larger objects. 

the order: @plbenveniste why would you not consider `"grey_matter": [2,3]` -- assuming that a lesion could span parts of GM and WM? 

```
"labels": {
    "background": 0,
    "spinal_cord": [1, 2, 3],
    "grey_matter": [2, 3],
    "lesions": [3]
},
"regions_class_order": [1, 2, 3],
```

**Comment by jcohenadad** [comment]
I'm not following your logic. Given that a lesion could be in the white matter, how could "grey_matter" be defined with [2, 3]? More context at https://github.com/ivadomed/utilities/issues/57#issuecomment-2998008890

**Comment by naga-karthik** [comment]
> But in this case, how do you define GM? 

okay, I get the problem now, it is a bit more complex than I imagined 😅 -- using the approaches mentioned by me and PL in https://github.com/ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star/issues/2#issuecomment-2997777155, we cannot get the fractional volume of lesion in the GM.

**Comment by jcohenadad** [comment]
in this case, coming back to https://github.com/ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star/issues/2#issue-3168385180 would you still recommend option 2 with region-based approach?

**Comment by naga-karthik** [comment]
> If, e.g., we want to quantify the fractional volume of a lesion within the GM, we need to know where it is.

I thought about it a bit more and we can still compute the fractional volume of a lesion within the GM. Consider the following diagram (sorry, it's a rough sketch) with the lesion partially in GM and WM. If we specify this order: `"grey_matter": [2, 3]` and `"lesion": [3]`, we get the following segmentation from the model

<img width="539" alt="Image" src="https://github.com/user-attachments/assets/9d5c3c36-eeea-4958-8051-0e32a123b9c7" />

(i.e. because we are including _both_ labels 2 and 3 for the GM in the `dataset.json` file, the model will learn to segment the GM fully. 

To get the fractional volume, once we get the output segmentation, we do 

```python
gm_seg = np.where(pred == 2, 1, 0)     # to get only the green part in the above figure, 
lesion_seg = np.where(pred == 3, 1, 0) # to get the lesion only (red part). 

# take intersection of both individual masks
lesion_inside_gm = gm_seg & lesion_seg

# To get fractional volume:
# 1. compute volume of full lesion (i.e vol. lesion_seg)
# 2. compute volume of lesion_inside_gm (i.e.  vol. lesion_inside_gm)
# 3. take ratio of 1 and 2. 

``` 

To answer your question 
> would you still recommend option 2 with region-based approach?

yes, using order described in https://github.com/ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star/issues/2#issuecomment-2997777155

**Comment by jcohenadad** [comment]
Thank you for providing more details, but the volume fraction was just an example— my point is that we _need_ the GM segmentation (eg to compute atrophy, to register it to the PAM50, etc.). So your example above would not work because a label=3 could either be in the GM or in the WM.

**Comment by plbenveniste** [comment]
From what I see, the explored solution with regionBased has a problem since nnUnet cannot provide classes separately. Using the following instructions would prevent us from having a clear segmentation of WM or GM since parts would be covered by lesions and therefore belonging to both WM and GM.  → This should not be done
```console
##### DON'T DO THIS ####
"labels": {
    "background": 0,
    "SC": [1, 2, 3],
    "WM": [1, 3],
    "GM": [2, 3],
    "lesions": [3]
},
"regions_class_order": [1, 2, 3],
##### DON'T DO THIS ####
```


Then, the only way to perform regionBased segmentation is to divide the lesions for WM and GM: it would look like this:
```console
"labels": {
    "background": 0,
    "WM": [1, 2],
    "GM": [3, 4],
    "lesions": [2,4]
},
"regions_class_order": [1, 3, 2, 4],
```

However, I am not sure if the algorithm will understand the splitting of the lesions depending on their belonging to WM or GM as the task is already complex enough. 
This could be tried. However, I would recommend a two step approach.
Step 1: region based segmentation of WM and GM:
```console
"labels": {
    "background": 0,
    "SC": [1, 2],
    "GM": [1],
    "WM": [2]
},
"regions_class_order": [1, 2],
```
Step 2: region based segmentation of lesions
```console
"labels": {
    "background": 0,
    "SC": [1, 2],
    "lesions": [2]
},
"regions_class_order": [1, 2],
```

**Comment by plbenveniste** [comment]
> About the 2-step approach: what do you mean exactly by that: 2 independently trained models? Or really a 2-step approach, where model 2 uses information from model 1? In the latter case, how would you do that?

For the 2-step approach, I don't have any strategy in mind. I was thinking of 2 independently trained model.

**Comment by jcohenadad** [comment]
Copying @Luugaaa 's comment (originally from https://github.com/neuropoly/idea-projects/issues/44#issuecomment-3075659973):

> I'd like to discuss the label processing [@jcohenadad](https://github.com/jcohenadad) [@AcastaPaloma](https://github.com/AcastaPaloma). For context, I've never had the opportunity to work on segmentation of overlapping classes before. Usually, a cat is a cat, and a dog is a dog. A pixel never really belongs to both classes at the same time. Because of that, I had to do some research about how region-based training works and how it's handled by nnunet.
> 
> Now, if my understanding is correct, to use region-based training, a `regions` tag is supposed to be present in this file `processed_data/nnunet_preprocessed/Dataset503_MagPhaseExp/dataset.json`. It gives nnunet the 3 regions (SC, GM, and lesion) and then nnunet will, on its own, understand how they overlap and how to handle this overlap through the label NIFTI we give it. (Again,) if my understanding is correct, during training, this process makes nnunet ask "is it GM?", "is it SC?" and "is it lesion?" separately. This is simulating multi-affiliation of a voxel --> nnunet understands what the whole lesion object is and what the whole GM object is; and then handles the overlap.
> 
> In the case of the `combine_labels.py` in the repo, we manually handle the overlap before giving the data to nnunet. Because of that, we're not training on regions but on multi-classes seemingly unrelated to each other. --> nnunet needs to understand what a lesion_only is, what a lesion_with_GM is, and what a GM_only is, as if they were completely unrelated to each other. This is, I believe, far more difficult than what the region-based training does. (side note: I'm pretty sure the set of NIFTI files doesn’t actually change; they are multi-class. The region-based training is just about the metadata we give to nnunet.)
> 
> Please tell me what you think about this.
> 
> Because of this questioning I had, I ended up rewriting parts of the preprocessing code for training. It's a bit messy right now, but I'll push it ASAP, unless you tell me I'm misunderstanding something. In that case, I'll stop working on the region-based training and I'll use the multi-class training (original code).

The benefits of hierarchical classes is well described on [nnUNet's repos](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/region_based_training.md#what-can-nnu-net-do), although I agree that the concept is still elusive to me. It would be good to compare various approaches. Also happy to discuss IRL-- it might be easier.

**Comment by Luugaaa** [comment]
Yes let's discuss it, @jcohenadad feel free to add a meeting in my calendar at your convenience, I'm currently updating my progress report 
Some preliminary results of the region based training on the validation set. The bottom right image being the final output.
Cyan --> SC
Red --> GM on SC
Purple --> Lesion on GM on SC
Magenta --> Lesion on SC

<img width="1130" height="731" alt="Image" src="https://github.com/user-attachments/assets/41d5716a-decb-4d7f-abf6-87c8d341c7f5" />

<img width="1130" height="731" alt="Image" src="https://github.com/user-attachments/assets/6d69228e-1305-4f4c-959e-209425a127ca" />

The raw prediction (bottom middle image) looks very bad because the model is not really confident, but the post processing (winner takes all and thresholding) show that it's still learning quite well. Still it's not good to have a model this uncertain and not robust, so I'm working on improving that right now. The main factors being (in my opinion) : 
* very high learning rate (0.01??)
* bad learning rate scheduler (over one training, the lr goes from 0.01 to 0.007...)
* lack of data
* poor augmentation

**Comment by Luugaaa** [comment]
For now, region based seems to be the best strategy when we compare [those results (multiclass)](https://github.com/ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star/issues/2#issuecomment-3005066836) with [these ones (region)](https://github.com/ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star/issues/2#issuecomment-3079128586). We'll keep working with that @AcastaPaloma

### Region Strategy 1
The first try with the regions was to use the regions `SC`, `GM` and `lesion`. This creates many unnecessary classes as GM always overlaps with SC, meaning that lesions that overlap with GM also necessarily overlap with SC etc... This imply having the following json data, in which the classes `GM`, `lesion` and `GM_lesion` are not actually used.

```json
{
    "labels": {
        "background": 0,
        "SC": 1,
        "GM": 2,
        "SC_GM": 3,
        "lesion": 4,
        "SC_lesion": 5,
        "GM_lesion": 6,
        "SC_GM_lesion": 7
    },
    "regions": {
        "SC": [
            1,
            3,
            5,
            7
        ],
        "GM": [
            2,
            3,
            6,
            7
        ],
        "lesion": [
            4,
            5,
            6,
            7
        ]
    }
}
```

At epoch 15, it produces results like this (below). Towards the end of the training, the post-processed results do get better, but the bad quality of the raw predictions is an indication the model has some difficulties and is not that confident in its predictions. 

<img width="1086" height="357" alt="Image" src="https://github.com/user-attachments/assets/93c40418-3bf1-4c99-bc37-ed0977a89736" />


### Region Strategy 2
On the other hand, when considering the WM instead of the SC, the json is cleaner and all classes are used.

```json
    "labels": {
        "background": 0,
        "WM": 1,
        "GM": 2,
        "lesion_WM": 3,
        "lesion_GM": 4
    },
    "regions": {
        "WM": [
            1,
            3
        ],
        "GM": [
            2,
            4
        ],
        "lesion": [
            3,
            4
        ]
    }
```

At epoch 15 the model produces results that are already sign a good leaning process and show a lot confidence (although it still needs to improve) (your intuition was spot on @jcohenadad !)

<img width="1086" height="357" alt="Image" src="https://github.com/user-attachments/assets/fb680203-bb29-4ef4-976b-b1c1f2bacd77" />

### Analysis
This difference is most likely due to the great number of classes in strategy 1 and the great dataset unbalance they create. The metrics also show how they perturbed the training. 
E.g : Class 2

<img width="1376" height="649" alt="Image" src="https://github.com/user-attachments/assets/530ad743-87f6-414c-a081-dbb89b48c070" />

*Note : Those classes actually exist, that is why the model is trying to predict them. They appear when, for example, a lesion is labeled a few pixel outside of the SC due a simple imprecision.*

### Conclusion
**I believe we can conclude that we should work with a region based training, using the minimal number of classes.**

**Comment by jcohenadad** [comment]
This is awesome @Luugaaa! One nitpick, I suggest to rename:
```json
    "labels": {
        "background": 0,
        "WM": 1,
        "GM": 2,
        "lesion_WM": 3,
        "lesion_GM": 4
    },
```
as: 
```json
    "labels": {
        "background": 0,
        "WM_nolesion": 1,
        "GM_nolesion": 2,
        "WM_lesion": 3,
        "GM_lesion": 4
    },
```
to avoid confusion (someone seeing a label called "WM" would not intuitively think that this refers to WM tissue _without_ lesion).
